### ISSUE TITLE
The dotless compiler is not a suitable replacement for the official lessc compiler

### ISSUE THREAD

**Issue by rjgotten** [madskristensen/WebCompiler#37]
Moving to dotless means a step back in functionality compared to what was offered with Web Essentials and its integration with NodeJS:
- No support for detached rulesets
- No support for canon JS-based plugin architecture
- No support for inline JS expressions
- No support for inline JS plugin references (2.5.0 `@plugin` keyword)
- Incomplete, buggy support for the `:extend` feature
- Partial implementation of the default function library
- On functions that *are* implemented; output that is not compliant with the official Less compiler
- etc.

The dotless  compiler is (on some fronts quite literally) years behind the curve on language features compared to the official Less compiler and switching the WebCompiler package to it is a **BAD IDEA**.

This will lead to headaches with third-party frameworks like Bootstrap, which base themselves on the state of the official compiler. This will lead to user confusion between usage of dotless by the WebCompiler package and usage of lessc with the grunt/gulp/etc. asset pipeline in VS2015 proper.

It is also a kick to the groin for those that have used and become accustomed to the features of the *real* compiler with WebEssentials 2013; have written nice and clean Less files making use of new language constructs and that now have to either give up on their compilation and preview tooling with VS 2015 or have to face stripping down their code to work with the subpar dotless implementation.

**Comment by rubenprins** [comment]
I concur. We switched all of our (VS/WE2013) projects to Web Essentials .less compilation because of dotless issues, and WE2013 providing real .less support. 

Forcing us back to dotless will mean that this extension is pretty much useless to us, since we do need an up-to-date (real/2.x) .less compiler.

I had hoped that this extension would solve WE2015's compiler regressions, but alas. Dotless simply can't keep up with the real Less compiler, and that's a deal breaker to us. Since we actively use Less 2.x features and frameworks like Bootstrap, this means most of our .less cannot even be compiled anymore.

In other words, we will need grunt/gulp files all over the place, with hand coded paths (whee!) And don't get me started on crashing watch tasks that you need to restart manually (**if* you've noticed they've crashed), and I still have no clue how to restart the watch task properly when the grunt/gulp file changes. It's giving me a headache.

And that's *still* better than dotless (for us).

**Comment by madskristensen** [comment]
It sounds like I'll have to change the implementation to real LESS compiler. I'll do that after I get home from traveling

**Comment by rubenprins** [comment]
That would be awesome. Thanks.

**Comment by madskristensen** [comment]
I'm having issues with this. How do I execute the official LESS compiler without relying on npm modules and node.exe?

**Comment by madskristensen** [comment]
...actually, I'm fine executing the script using node.exe, but it will be problematic to involve npm and `node_modules`. If I can just find and execute a single .js file using node.exe then that'll be fine

**Comment by ashrafsabrym** [comment]
dotLess should give better performance, but unfortunately, it doesn't cover a lot of language features.
I have problems compiling my files with Web Compiler too.

**Comment by rubenprins** [comment]
@madskristensen, Is there a technical reason for not just bundling the `node_modules` folder npm creaties on the build server? Node.exe doesn't care how its filled, as long as the required modules are there.

Otherwise, maybe Browserify with the `--node` option might be a starting point. As long as there are no native modules or non-require'd files involved, this might work too.

**Comment by rjgotten** [comment]
> @ashrafsabrym :
> dotLess should give better performance

I would not be so sure about that.

I tried to use dotless about a year ago to compile Less to CSS on-the-fly using a handler. At one point I could spend up to 7 minutes (yes; *minutes*) waiting for my website's Less to compile down to CSS.

Eventually I replaced the whole thing with a process call to node.exe and the official lessc compiler, which reduced the compilation time down to 2 seconds on the very same files.


> @rubenprins :
> Otherwise, maybe Browserify with the `--node` option might be a starting point. As long as there are no native modules or non-require'd files involved, this might work too.

Less scans a few pre-determined paths for plugins and pulls them in via `require` calls with hard paths instead of relying purely on the lookup logic that is internal to `require`. The mime packages on which it depends also load some files based on the 'current module path'. So that's a no-go, I'm afraid.

Using the pre-browserified version of lessc directly is possibly even worse, since it disables built-in features like the `data-uri()` function from its standard functions library and relies on DOM constructs like `XmlHttpRequest` that you'd have to shim. Web Essentials 2012 went down that road with the Windows Scripting Host: it is **not** pretty...


@madskristensen 
Let me guess; the problem you have with npm is with path depth of nested `node_modules` folders and the fact that it does not agree with the Win32 APIs TFS uses for file access?

WebEssentials 2013 works around that part by bundling a pre-flattened tree of modules, but it prevents upgrades and really bloats the package, so you probably don't want to do that. You want to use the new NPM/Gulp support to pull down the necessary modules when the extension is installed into Visual Studio or when it installs some form of MSBuild / TFSBuild support into a project.

If I'm correct sofar; there's actually a dead easy way to script NPM installation and use a package available on NPM to perform the flattening of the modules. I published how to do that as a hotfix for the broken `node_modules` folder in WebEssentials 2013 v 2.6.12: https://github.com/madskristensen/WebEssentials2013/issues/1917

**Comment by rubenprins** [comment]
@rjgotten, I just performed a simple test with
```
browserify --node lessc > bundled-lessc.js
```
(Note that I'm compiling the `bin/lessc`, not `index.js`.)

This does give me a working script I can point node.exe to. The node intrinsics are still there, not patched by browser-shims or using `XmlHttpRequest`.

I admit, it was just a crude test I performed, but the basics appeared to work. And it still supports stuff like `@imports` and `data-uri`, and gives me no MIME-related errors (Browserify is smart enough to include the required .json files). And it reduces the less compiler to a single file.

You are correct that running the `lessc` compiler though Browserify this way will make it harder to support plugins (I didn't look that far). At the very least you would have to patch `lessc` to `require` the plugins upfront, and adjust the plugin loader code accordingly.

Your other concerns in madskristensen/WebEssentials2013#1917 remain valid though, and the real lessc compiler would be preferable (and more serviceable).

**Comment by madskristensen** [comment]
The reason I'm reluctant to use node.exe and the npm modules of `lessc` is due to the increase in moving parts needed. I've had very bad experiences using this in Web Essentials 2013 and I don't want to repeat that again.

However, where I am happy using node.exe is in the case when node.exe and the npm modules needed can be included in the extension without every single file in the node_modules folder is listed in the .csproj file. Also, the packages must be flattened (but that's easy as earlier mentioned). It's important that the execution is completely isolated so that the user won't have to install node.js on the box/CI server.

One such way of doing that is what @duncansmart is doing here https://github.com/duncansmart/LessCoffee which is basically to flatten and zip up the node modules and then unzip them on first use. I like that idea.

Also, it's faster to compile LESS/CoffeeScript if we're spinning up a node server that is kept alive as long as VS is open. All of this sounds ok to me except for the additional mechanical parts which increases the risk of failure and makes it harder to maintain, but seems necessary.

So, the system would do this:

1. Isolate node.exe and node_modules inside the extension directory
  - Download latest version of node.exe and npm when missing on disk
  - Install the node_modules using package.json and npm when missing on disk
2. Zip up the node_modules and node.exe and embed in WebCompiler assembly
3. Unzip the node_modules on first use
4. Spin up a node server to handle both LESS, CoffeeScript and Iced CoffeeScript
5. Make this as simple as possible so it easy to maintain
6. Design it in a generic way so compilers for other languages can be added easily

I'm not  very good with node and npm so I could really use some help to get this started

**Comment by rjgotten** [comment]
In general the approach is good and more or less similar to Web Essentials 2013.

One thing springs to mind though: how would you handle Less plugins like autoprefixer or clean-css?

When called via the commandline, lessc creates an instance of the main compiler class and then scans for and adds plugins based on passed arguments. When you access the compiler in code, it becomes your own responsibility to initialize the plugins correctly and add them to the compiler: http://lesscss.org/usage/#plugins-using-a-plugin-in-code

You can still keep around a single instance of the Less compiler (thank goodness), but you *will* need a way to specify those plugins... That may be a setting/option in the Visual Studio setting, but would in turn also require that you `npm install` the necessary modules depending on which plugins are configured.

Unless ofcourse; you're only going to support a limited set of well-known ones that are bundled in the `node_modules` extracted by the extension on first use.

**Comment by madskristensen** [comment]
I don't want this to be pluggable to the extend you mention. Instead, hand pick a few plugins to support but nothing more. People should use Gulp/Grunt if they want full granularity.

**Comment by madskristensen** [comment]
Please help test the latest build with the official LESS compiler running in node.js by downloading it here http://vsixgallery.com/extension/148ffa77-d70a-407f-892b-9ee542346862/

**Comment by BryanTheCrow** [comment]
Thanks. Fixed my compilation issues. I had to disable and re-enable compile-on-build though (not sure if that's because I didn't uninstall the previous version first).

**Comment by kimsmith999** [comment]
I am not a UX designer and not well versed in C# or VS but have a real mess in VS 2013 awith the .less files and am having issues with compilers.  
http://localhost:55876/Infrastructure/MVC/BundleTransforms/DotLess/LessBundle.c
http://localhost:55876/App_Start/BundleConfig.css
I have installed Web Essentials 2013, node.js, and grunt.  I am trying to eliminate the old dotless references.  I deleted the dotless reference but still see it in the code.  Can you give me any pointers on what my bundles file should include to use the correct compiler and get rid of dotless?
